[red-knot] Infer target types for unpacked tuple assignment

dhruvmanila opened the pull request:
## Summary

This PR adds support for unpacking tuple expression in an assignment statement where the target expression can be a tuple or a list (the allowed sequence targets).

The implementation introduces a new `infer_assignment_target` which can then be used for other targets like the ones in for loops as well. This delegates it to the `infer_definition`. The final implementation uses a recursive function that visits the target expression in source order and compares the variable node that corresponds to the definition. At the same time, it keeps track of where it is on the assignment value type.

The logic also accounts for the number of elements on both sides such that it matches even if there's a gap in between. For example, if there's a starred expression like `(a, *b, c) = (1, 2, 3)`, then the type of `a` will be `Literal[1]` and the type of `b` will be `Literal[2]`.

There are a couple of follow-ups that can be done:
* Use this logic for other target positions like `for` loop
* Add diagnostics for mis-match length between LHS and RHS

## Test Plan

Add various test cases using the new markdown test framework.
Validate that existing test cases pass.

dhruvmanila commented:
There's some more work needs to be done, please don't look ;)

dhruvmanila commented:
This is completely an incorrect solution, I quickly realize that this will require structural matching between the target expression and the value type. It would also be good to add caching here because we've definitions for each name node on the LHS and so we'd perform structural matching against the same target and value type for each name node that's present on the LHS.

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/unpacking.md:
````diff
@@ -0,0 +1,149 @@
+
+### Starred expression (1)
+
+TODO: The error message here is because `infer_starred_expression` isn't complete
+
+```py
+[a, *b, c, d] = (1, 2)  # error: "Object of type `None` is not iterable"
````
As mentioned, the implementation of `infer_starred_expression` adds this diagnostic. This should be fixed when implementing assignment to a starred expression.

dhruvmanila replied:
Yeah, I can make the TODO clearer here.

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/unpacking.md:
````diff
@@ -0,0 +1,149 @@
+## String
+
+### Simple unpacking
+
+```py
+a, b = 'ab'
+reveal_type(a)  # revealed: LiteralString
+reveal_type(b)  # revealed: LiteralString
+```
````
It's interesting that mypy doesn't support string unpacking (https://github.com/python/mypy/issues/13823) while Pyright can.

dhruvmanila commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,133 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                    .elements(builder.db)
+                                    .get(starred_index..=end_index)
+                                    .unwrap();
+                                // TODO: Combine the types into a list type.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
```
This is basically doing something like combine `Literal[1], Literal[2]` into `int` and the assignment to starred expression should add the `list` to make it `list[int]`.

dhruvmanila commented on crates/red_knot_workspace/tests/check.rs:
```diff
@@ -65,6 +65,17 @@ impl<'db> PullTypesVisitor<'db> {
         }
     }
+
+    fn visit_assign_target(&mut self, target: &Expr) {
+        match target {
+            Expr::List(ast::ExprList { elts, .. }) | Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                for element in elts {
+                    self.visit_assign_target(element);
+                }
+            }
+            _ => self.visit_expr(target),
+        }
+    }
```
The reason this is required is because in a code like `(a, b) = (1, 2)`, the LHS expression is not available.

carljm replied:
Since the RHS is a standalone expression whose type should be a cached Salsa query, I think we _could_ query the type of that expression one more time and set a type on the entire LHS expression. But I'm not sure there's any value to doing that; I'm happy with this change to our coverage goal.

dhruvmanila replied:
Yeah, I agree. As you've mentioned, I'm also unsure of what the benefit could be of getting the type of the LHS expression which is either a tuple or list.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
+                                );
+
+                                // E.g., in `(a, *b, c, d) = ...`, the index of starred element `b`
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
```
This doesn't look right to me; the `- 2` seems hardcoded to the case where `starred_index == 1`, and all the starred unpacking tests also have `starred_index == 1`. If I change one of the tests to have the starred element in a different position, inference of the element types from the end doesn't seem to work correctly.

If this variable is supposed to be the number of remaining elements in the unpacking target, it seems like that should be something like `elts.len() - (starred_index + 1)` instead?

And let's make sure we add a test for starred unpack where the starred element is not in second position.

dhruvmanila replied:
Yes, thanks for catching that, I totally did not intend that to happen.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                // E.g., in `(a, *b, c, d) = ...`, the index of starred element `b`
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
+                                let end_index =
```
I don't think the name `end_index` is very clear here. Seems like it should be the index of the last element to be "gobbled" by the starred element; maybe `starred_end_index`?

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
+                                let end_index =
+                                    tuple_ty.len(builder.db) - starred_index - remaining;
```
If we calculate `remaining` correctly above, then I think `end_index` should be purely a function of `tuple_ty.len()` and `remaining`; in fact it seems like it should be `tuple_ty.len(builder.db) - remaining - 1`?

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                    .unwrap();
+                                // TODO: Combine the types into a list type.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
```
Somewhere this Todo gets swallowed and turned into Unknown. Not sure why, and it's not really a big deal, but if it's not hard to preserve it as a Todo that would be nice.

dhruvmanila replied:
I got it, it's because the `inner` function does not account for `Expr::Starred` and only checks for `Expr::Name`. So, it would return `None` which then defaults to `Unknown`.

So, the diff would become
```diff
                 ast::Expr::Name(name) if name == variable => {
                     return Some(value_ty);
                 }
+                ast::Expr::Starred(ast::ExprStarred { value, .. }) => {
+                    // TODO: Wrap the `value_ty` in a list type.
+                    return inner(builder, value, value_ty, variable);
+                }
                 ast::Expr::List(ast::ExprList { elts, .. })
                 | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
                     Type::Tuple(tuple_ty) => {
                         let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
```

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                    tuple_ty.elements(builder.db).get(end_index + 1..).unwrap(),
+                                );
+                                Cow::Owned(element_types)
+                            } else if tuple_ty.len(builder.db) == elts.len() - 1 {
+                                let mut element_types = tuple_ty.elements(builder.db).to_vec();
+                                element_types.insert(starred_index, Type::Todo);
+                                Cow::Owned(element_types)
```
This is handling the case where the starred element would be bound to nothing, i.e. match zero elements and turn into an empty list? Maybe add a comment with an example of what it is handling.

I wonder if this case could just be integrated into the above case; we should be able to detect that `starred_index..=end_index` is an empty range and just handle that as needed (or just handle an empty `starred_element_types`.)

dhruvmanila replied:
> I wonder if this case could just be integrated into the above case; we should be able to detect that `starred_index..=end_index` is an empty range and just handle that as needed (or just handle an empty `starred_element_types`.)

Yeah, I think this is possible and should make it simpler. Thanks for the suggestion.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                            }
+                        }
+                    }
+                    Type::StringLiteral(string_literal_ty) => {
+                        for (index, element) in elts.iter().enumerate() {
+                            let string_length = string_literal_ty.len(builder.db);
+                            let element_ty = if index < string_length {
+                                Type::LiteralString
+                            } else {
+                                Type::Unknown
+                            };
+                            if let Some(ty) = inner(builder, element, element_ty, variable) {
+                                return Some(ty);
+                            }
+                        }
+                    }
```
It seems like to assign Unknown correctly here (and in the future to correctly issue diagnostics for mismatched unpacking) we also need to account for starred element, if present?

Alternatively, we could just generalize the tuple RHS case above, so for a string RHS we get a slice of one-character StringLiteral types and otherwise all the logic is identical? That would avoid needing to implement the starred index handling twice.

I don't think it's necessary that we precisely infer literal string types from unpacking a literal string, only suggesting that in case it is just as easy as handling starred element explicitly here?

dhruvmanila replied:
Hmm, it might be useful to generalize it. We can convert the `StringLiteral` type into a `TupleType` containing `n` number of either `LiteralString` or `StringLiteral` (with the corresponding character) where `n` is the length of the containing string.

dhruvmanila replied:
I've addressed this feedback by splitting the `StringLiteral` type into a tuple type containing equal amount of `LiteralString` and delegating it back to the `inner` function for correct handling of starred expression.

carljm reviewed (changes requested):
This is great! A few issues, but seems quite close. Make sure for any bugs you fix in updating the PR, you also add a new test that would fail without that fix.

dhruvmanila commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -292,6 +292,13 @@ impl<'db> Type<'db> {
         matches!(self, Type::Never)
     }
 
+    pub const fn as_tuple_type(&self) -> Option<&TupleType<'db>> {
```
I think the reason those method names are using the `into_` prefix is because they consume `self` and return the relevant type while this method only returns a reference.

Also, this method isn't required now but was in an earlier version. Sorry for the churn, I'll remove this.

AlexWaygood replied:
Yes, the `into_*` naming convention is because it consumes `self` -- but because `Type` is `Copy`, consuming `self` and returning an owned version of the underlying data is actually more ergonomic (and probably no less performant, or at least not significantly so) for these methods than returning a reference to the underlying data

dhruvmanila commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1151,13 +1154,23 @@ impl<'db> TypeInferenceBuilder<'db> {
+                }
+            }
+            _ => {
+                // TODO: Remove this once we handle all possible assignment targets.
```
Yeah, I think that makes sense.

I'm just trying to understand for "assignment that doesn't create a Definition." do you mean that something like `a.b = 1` won't create a `Definition` or it won't create an `AssignmentDefinition`?

carljm replied:
I mean that `a.b = 1` (and `a[b] = 1`) won't create a `Definition` at all, since it doesn't define any name in the local scope. Though ultimately it may create some other kind of struct that we track relative to control-flow, if we want to narrow types based on such assignments.

carljm replied:
It's not important that we do anything about this comment in this PR, the existing TODO is OK, we have more to figure out here regardless.

dhruvmanila commented on crates/red_knot_python_semantic/src/semantic_index/definition.rs:
```diff
@@ -484,18 +493,35 @@ impl ImportFromDefinitionKind {
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum AssignmentKind {
+    Sequence,
+    Other,
```
Yeah, that's correct. This helped me uncover https://github.com/astral-sh/ruff/issues/13759

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/unpacking.md:
```diff
@@ -0,0 +1,149 @@
+
+### Non-iterable unpacking
+
+TODO: Remove duplicate diagnostics. This is happening because for a sequence-like
+assignment target, multiple definitions are created and the inference engine runs
+on each of them which results in duplicate diagnostics.
```
Yeah, that's correct, we'd need a way to do unpacking once and cache it for future lookup. I can take this up as a follow-up at a high priority, will need to think about the required changes.

dhruvmanila commented:
@carljm Thanks for the great review, I've addressed all of the feedback except for the duplicate diagnostic issue (https://github.com/astral-sh/ruff/pull/13316#discussion_r1800219731) which I'll fix as a follow-up.

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/unpacking.md:
````diff
@@ -0,0 +1,235 @@
+reveal_type(d)  # revealed: Literal[6]
+```
+
+### Starred expression (5)
+
+```py
+# TODO: Remove 'not-iterable' diagnostic
+[a, b, *c] = (1, 2, 3, 4)  # error: "Object of type `None` is not iterable"
+reveal_type(a)  # revealed: Literal[1]
+reveal_type(b)  # revealed: Literal[2]
+# TODO: Should be List[int] once support for assigning to starred expression is added
+reveal_type(c)  # revealed: @Todo
+```
````
The new test case where the starred expression isn't at the same position as other test cases.

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/unpacking.md:
````diff
@@ -0,0 +1,235 @@
+reveal_type(b)  # revealed: LiteralString
+```
+
+### Starred expression (1)
````
These test cases are similar to that of the `Tuple` type above and it only differs such that we use strings and the correct revealed type.

carljm reviewed (commented):
Changes look great! I noticed a few more things in reviewing just now, sorry I didn't catch these yesterday!

dhruvmanila commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                return Some(ty);
+                            }
+                        }
+                    }
```
Yeah, that makes sense. I've updated the test cases to include this at both top level and nested place. This also removes the need to do it in `infer_assignment_definition`. Thanks!

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/definition.rs:
```diff
@@ -224,12 +226,17 @@ impl DefinitionNodeRef<'_> {
+                target,
+                variable,
+                kind,
+            }) => DefinitionKind::Assignment(AssignmentDefinitionKind {
+                value: AstNodeRef::new(parsed.clone(), value),
+                target: AstNodeRef::new(parsed.clone(), target),
+                variable: AstNodeRef::new(parsed, variable),
```
Creating `AstNodeRef`'s isn't expensive but it also isn't free because it requires bumping an `Arc` (and decrementing it later). 

It might be worth to keep storing the assignment only with an index for the right target. I'm not sure what `variable` is for.

dhruvmanila replied:
> It might be worth to keep storing the assignment only with an index for the right target. I'm not sure what `variable` is for.

(lol, after writing a couple of sentences I realize which target you're referring to.)

Ah, yes, I think that can be done.

The `variable` is used to find the symbol the definition belongs to in case the LHS is a sequence. It is also used to add the "symbol -> type" mapping in `self.types.expressions` hashmap.

carljm replied:
Using indices is quite tricky because of arbitrary nesting of unpacking; this isn't like imports where it's a flat list of items. It's possible but requires keeping a counter across the entire recursive traversal; I think the current code is a lot clearer.

I was unclear about `variable` at first also; I think it is needed for the unique per-Definition NodeKey. Perhaps `name` would be clearer than `variable` -- I think it is always a `Name` node.

dhruvmanila replied:
> Using indices is quite tricky because of arbitrary nesting of unpacking; this isn't like imports where it's a flat list of items. It's possible but requires keeping a counter across the entire recursive traversal; I think the current code is a lot clearer.

I _think_ Micha is talking about the `targets` list in a multi-assignment statement like `x = y = 1` in which case we could just store the single `AstNodeRef<ast::StmtAssign>` and a `usize` index into the `targets` list. @MichaReiser can confirm though

> I was unclear about `variable` at first also; I think it is needed for the unique per-Definition NodeKey. Perhaps `name` would be clearer than `variable` -- I think it is always a `Name` node.

Yes, the `name` field already existed by the name of `target`, I mainly renamed it to variable to avoid any clash with the `targets` field in the node itself. I think `name` is more clear.

MichaReiser commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                let mut element_types = Vec::with_capacity(elts.len());
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
```
```suggestion
                                    &tuple_ty.elements(builder.db)[..starred_index],
```

Possibly, depending on what `elements` returns

dhruvmanila commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                    return Some(value_ty);
+                }
+                ast::Expr::Starred(ast::ExprStarred { value, .. }) => {
+                    // TODO: Wrap the `value_ty` in a list type.
+                    return inner(builder, value, Type::Todo, variable);
```
Yeah, I realized that while resolving your previous comment and playing around with some test cases. This makes sense, thanks.

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                        }
+                    }
+                    Type::StringLiteral(string_literal_ty) => {
+                        // Deconstruct the string literal to delegate the inference back to the
+                        // tuple type for correct handling of starred expressions.
+                        let value_ty = Type::Tuple(TupleType::new(
+                            builder.db,
+                            vec![Type::LiteralString; string_literal_ty.len(builder.db)]
+                                .into_boxed_slice(),
+                        ));
```
I think ideally, for something like this:

```py
a, (b,), c, d = "abcd"
```

We'd infer `a` as being `Literal["a"]`, `b` as `Literal["b"]`, `c` as `Literal["c"]` and `d` as `Literal["d"]`. (At the moment they're all `LiteralString`.)

dhruvmanila replied:
Yeah, I think that depends on how precise we want the types. For this case, I think it might make sense to do so.

dhruvmanila replied:
I can do that as a follow-up.

AlexWaygood replied:
Sounds good!

carljm replied:
I agree that we _can_ do this, I'm not sure we ever _need_ to do it (pyright doesn't do it, it will have some cost in complexity and perf), and I'm pretty sure there's no need to do it _now_. I would suggest we defer it until or unless we see real motivating use cases. Avoiding duplicate diagnostics, supporting unpacking in for loops, and handling starred elements correctly are all much more important follow-ups. (Though handling starred elements correctly is blocked on generic list type, I think.)

AlexWaygood replied:
fine by me if we don't do it, that all makes sense. but in that case I think we should leave a comment saying that it's a deliberate choice to infer a less precise type -- it _looked_ like it could have been a mistake to me when I was reading through the diff :-)

carljm replied:
Makes sense, I'll add such a comment!

AlexWaygood reviewed (commented):
Nice -- this is complex stuff to get right!! A couple of small points on top of Micha's and Carl's comments:

dhruvmanila commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                );
+                                Cow::Owned(element_types)
+                            } else {
+                                Cow::Borrowed(tuple_ty.elements(builder.db).as_ref())
```
We discussed this internally and agreed that this is simpler / easier compared to the other alternative where we'd need to convert the type corresponding to the starred element at two places.

dhruvmanila commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -566,8 +566,19 @@ where
+                        ast::Expr::Name(_) => Some(AssignmentKind::Name),
+                        _ => None,
+                    };
+                    if let Some(kind) = kind {
+                        self.current_assignment = Some(CurrentAssignment::Assign {
+                            target,
+                            value: &node.value,
+                            kind,
+                        });
+                    }
                     self.visit_expr(target);
                 }
                 self.current_assignment = None;
```
Ah yes, good catch, thanks!

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1166,18 +1179,135 @@ impl<'db> TypeInferenceBuilder<'db> {
+
+        let value_ty = self.expression_ty(value);
+
+        let target_ty = match (value_ty, kind) {
+            (_, AssignmentKind::Sequence) => self.infer_sequence_unpacking(target, value_ty, name),
+            _ => value_ty,
+        };
```
```suggestion
        let target_ty = match kind {
            AssignmentKind::Sequence => self.infer_sequence_unpacking(target, value_ty, name),
            _ => value_ty,
        };
```

carljm commented:
Given it's quite late now for @dhruvmanila , I'm going to apply my one suggested change and go ahead and merge this, to reduce our number of outstanding PRs and thus potential conflicts!

dhruvmanila commented:
Thanks @carljm for the reviews and merging!
